kernel/src/process/process_vm/mod.rs

```diff
@@ -64,109 +63,88 @@ use crate::{prelude::*, vm::vmar::Vmar};
  *  (low address)
  */
 
-/// The process user space virtual memory
+/// The process user space virtual memory.
 pub struct ProcessVm {
-    root_vmar: Mutex<Option<Vmar<Full>>>,
-    init_stack: InitStack,
-    heap: Heap,
-    #[cfg(target_arch = "riscv64")]
-    vdso_base: AtomicUsize,
-}
-
-/// A guard to the [`Vmar`] used by a process.
-///
-/// It is bound to a [`ProcessVm`] and can only be obtained from
-/// the [`ProcessVm::lock_root_vmar`] method.
-pub struct ProcessVmarGuard<'a> {
-    inner: MutexGuard<'a, Option<Vmar<Full>>>,
-}
-
-impl ProcessVmarGuard<'_> {
-    /// Unwraps and returns a reference to the process VMAR.
-    ///
-    /// # Panics
-    ///
-    /// This method will panic if the process has exited and its VMAR has been dropped.
-    pub fn unwrap(&self) -> &Vmar<Full> {
-        self.inner.as_ref().unwrap()
-    }
-
-    /// Returns a reference to the process VMAR if it exists.
-    ///
-    /// Returns `None` if the process has exited and its VMAR has been dropped.
-    pub fn as_ref(&self) -> Option<&Vmar<Full>> {
-        self.inner.as_ref()
-    }
-
-    /// Sets a new VMAR for the binding process.
-    ///
-    /// If the `new_vmar` is `None`, this method will remove the
-    /// current VMAR.
-    pub(super) fn set_vmar(&mut self, new_vmar: Option<Vmar<Full>>) {
-        *self.inner = new_vmar;
-    }
+    vmar: Option<Vmar<Full>>,
+    inner: Arc<ProcessVmInner>,
```

Comment:
In fact, the `Arc` in the two fields should be merged?

Reply:
> we could no longer have an InitStackReader that holds a reference to Vmar. 

I am not sure what do you mean by saying so?

#2449 will alter the way that `InitStackReader` holds a reference to `Vmar`. Will this help? I'm not sure what difficulties you're facing.